lib/src/sdk/dart.dart

```diff
@@ -31,6 +40,24 @@ class DartSdk extends Sdk {
     return aboveExecutable;
   }();
 
+  /// The loaded `sdk_packages.yaml` file if present.
+  static final SdkPackageConfig? _sdkPackages = () {
+    var path = p.join(_rootDirectory, 'sdk_packages.yaml');
```

Comment:
Could we put this in a sub directory?
I can see that the `version` file is at the top level, but nothing else seems to be there, and keeping a clean your it's usually a good thing.
If no current subdir works, we can add `cfg/`.

Reply:
How about pkg/?

It would fit well with the bin and lib directories inside the Dart SDK already, and match the pkg/ directory inside the Dart SDK's source code.

pkg/packages.yaml e.g.

Reply:
And then put the packages in there, like pkg/macros/

Reply:
That sounds fine to me, @jonasfj wdyt? I think this file was originally your idea :)

Reply:
I'd say it should be in the root.

But we haven't landed this in Flutter yet :)

But that's IMO where it'd make sense to put it in Flutter, since it needs to refer to both `packages/` and `bin/cache/pkg/`.

Arguably, I proposed this file because directory layout in Flutter is confusing and there is no build system that copies thing into the correct location when the Flutter SDK is built. Because as far as I understand it, you still don't build the SDK into artifact, you just let it modify itself.

We could choose not to have it for the Dart SDK, because it does have a build system. And the Dart SDK does have a concept of building into a binary artifact. And adding something to the resulting binary artifact is an explicit decision.

So it could be argued we don't need it, or we could write build rules that read it, and copy folders listed in there into the pkg/ folder in the output.

-----

For consistency with flutter, I'd suggest putting it in the root.

Reply:
> The main reason we want it for Flutter is because they don't have a build step so all their internal packages are SDK packages by accident more than intentional decision.

Just saying: we could also fix this by moving the internal packages out of the `packages/` folder... Sometimes low-tech is good.

I'm not strongly against an indirection file. I just don't see a big need.

Reply:
> Just saying: we could also fix this by moving the internal packages out of the packages/ folder... Sometimes low-tech is good.

True, we could try that.

> I'm not strongly against an indirection file. I just don't see a big need.

That's fair, and also what @sortie is arguing.
Especially, for the Dart SDK, we might not need it since there are build-rules that makes copying sdk-packages into the `out/` folder rather explicit (not something done by accident).

Reply:
While I agree in principle that it is not necessary (clearly, Flutter works today without it), I do think it is a much better design to have the SDK be in control of where its own packages live, instead of having pub hard code this knowledge. Why have tight coupling where we can easily avoid it?

I see also a `dartdoc_options.yaml` file in the root of the dart SDK, I don't think it hurts to have this file there as well.

We have a lot of this similar coupling in other places, even with packages published on pub, such as the locations of platform dill files and executables. It might be nice to move those into a config file as well at some point which would avoid a lot of this.